File: sorting/mergeSort/mergeSort.js
function merge(arr1, arr2) {
    let arr1Length = arr1.length;
    let arr2Length = arr2.length;
    let i = 0; let j = 0;
    let result = []
    while (i < arr1Length && j < arr2Length) {
        if (arr1[i] < arr2[j]) {
            result.push(arr1[i]);
            i++;
        } else {
            result.push(arr2[j]);
            j++;
        }
    }
    while (i < arr1Length) {
        result.push(arr1[i]);
        i++
    }

    while (j < arr2Length) {
        result.push(arr2[j])
        j++
    }

    return result;
}


function mergeSort(arr){
   if(arr.length<=1) return arr;
let mid = Math.floor(arr.length/2);
let left = mergeSort(arr.slice(0,mid));
let right= mergeSort(arr.slice(mid));
return merge(left,right)
}
console.log(mergeSort([1, 2, 5, 6,3,8,4,7]));

File: sorting/mergeSort/mergeSort.java
import java.sql.ResultSet;
import java.util.Arrays;

public class mergeSort {

    static int[] merge(int[] arr1, int[] arr2) {

        int i = 0;
        int j = 0;
        int[] result = new int[arr1.length + arr2.length];
        int k = 0;
        while (i < arr1.length && j < arr2.length) {
            if (arr1[i] < arr2[j]) {
                result[k++] = arr1[i];
                i++;
            } else {
                result[k++] = arr2[j];
                j++;
            }
        }

        while (i < arr1.length) {
            result[k++] = arr1[i];
            i++;
        }
        while (j < arr2.length) {
            result[k++] = arr2[j];
            j++;
        }
        return result;
    }

    static int[] mergeSortFn(int[] arr) {
        if (arr.length <= 1)
            return arr;

        int mid = arr.length / 2;
        int[] left = new int[mid];
        int[] right = new int[arr.length - mid];

        for (int k = 0; k < left.length; k++) {
            left[k] = arr[k];
        }
        for (int k = 0; k < right.length; k++) {
            right[k] = arr[k + mid];
        }

        int[] result = merge(left, right);
        return result;
    }

    public static void main(String[] args) {

        System.out.println(Arrays.toString(mergeSortFn(new int[] { 1, 2, 5, 6, 3, 8, 4, 7 })));
    }
}
